lib/state_change_syncer.go

```diff
@@ -0,0 +1,431 @@
+func NewStateChangeSyncer(desoParams *DeSoParams, stateChangeFilePath string) *StateChangeSyncer {
+	stateChangeIndexFilePath := fmt.Sprintf("%s-index", stateChangeFilePath)
+	stateChangeFile, err := openOrCreateLogFile(stateChangeFilePath)
+	if err != nil {
+		glog.Fatalf("Error opening stateChangeFile: %v", err)
+	}
+	stateChangeIndexFile, err := openOrCreateLogFile(stateChangeIndexFilePath)
+	if err != nil {
+		glog.Fatalf("Error opening stateChangeIndexFile: %v", err)
+	}
+	stateChangeFileInfo, err := stateChangeFile.Stat()
+	if err != nil {
+		glog.Fatalf("Error getting stateChangeFileInfo: %v", err)
+	}
+
+	return &StateChangeSyncer{
+		StateChangeFile:       stateChangeFile,
+		StateChangeIndexFile:  stateChangeIndexFile,
+		StateChangeFileSize:   uint32(stateChangeFileInfo.Size()),
+		DeSoParams:            desoParams,
+		UnflushedBytes:        make(map[uuid.UUID]UnflushedStateSyncerBytes),
+		StateSyncerMutex:      &sync.Mutex{},
+		ConnectedMempoolTxMap: make(map[BlockHash]*StateChangeEntry),
+	}
+}
+
+// Reset resets the state change syncer by truncating the state change file and index file.
+func (stateChangeSyncer *StateChangeSyncer) Reset() {
+	err := stateChangeSyncer.StateChangeFile.Truncate(0)
+	if err != nil {
+		glog.Fatalf("Error truncating stateChangeFile: %v", err)
+	}
+
+	err = stateChangeSyncer.StateChangeIndexFile.Truncate(0)
+	if err != nil {
+		glog.Fatalf("Error truncating stateChangeIndexFile: %v", err)
+	}
+
+	stateChangeSyncer.StateChangeFileSize = 0
+
+	stateChangeSyncer.EntryCount = 0
+}
+
+// _handleMempoolTransaction constructs a StateChangeEntry for a given mempool transaction. Upon connect, this entry
+// is added directly to the state change file. Upon disconnect, we look up the entry in the ConnectedMempoolTxMap
+// and add it to the bytes to be written to the state change file upon DB flush.
+func (stateChangeSyncer *StateChangeSyncer) _handleMempoolTransaction(event *MempoolTransactionEvent) {
+	// We shouldn't need this for mempool transactions, as they all occur on the same thread.
+	// Just adding for completeness.
+	stateChangeSyncer.StateSyncerMutex.Lock()
+	defer stateChangeSyncer.StateSyncerMutex.Unlock()
+
+	stateChangeEntry := event.StateChangeEntry
+	stateChangeEntry.IsMempoolTx = true
+
+	// If the transaction is connected, add it to the connected mempool map. Upon disconnect, we'll look this up and pass
+	// it to the state change file so the consumer can use them to revert the mempool transaction.
+	if event.IsConnected {
+		// Check to see if the operation in question has a "core_state" annotation for its key.
+		if !isCoreStateKey(stateChangeEntry.KeyBytes) {
+			return
+		}
+		// Create a copy of the state change entry for the map so that the utxoOps don't get deleted when they're removed
+		//from the stateChangeEntry struct (utxoOps are only included for disconnects).
+		mapStateChangeEntry := *stateChangeEntry
```

Comment:
i *think* this is probably a fine way to create a copy as long as we don't modify anything that references a pointer underneath.  A little concerned about UtxoOps below.